# Preprocessing

## Install snakePipes

Use the environment yaml `snakepipes.yaml` to install snakePipes v2.7.2 via `conda env create -f snakepipes.yml`. Alternatively install the most recent version following the tutorial on the respective [documentation](https://snakepipes.readthedocs.io/en/stable/content/setting_up.html#installing-conda).

## Workflow

1. Prepare your working directory with a folder for fastq files.
2. Make sure the fastq files are all within one folder and named properly (`_R1` and `_R2` files of each sample needed including the ENCODE reference).
3. Load the snakepipes conda environment with `conda activate snakepipes`
4. Execute the DNA-mapping pipeline using the following command (this will put all pipeline outputs into the `sp` subfolder):
```
DNA-mapping --DAG --trim --trimmerOptions `-a nexteraF=CTGTCTCTTATA -A nexteraR=CTGTCTCTTATA` --dedup --mapq 2 -j 20 -i ./fastq/ -o ./sp/ hg38
```
5. After the pipeline finishes, if you want to recreate the peak file from the 0.1%FA-cryo sample, run the ATAC-seq pipeline as follows (this uses the folder system generated by `DNA-mapping`):
```
ATAC-seq --DAG -d ./sp/ hg38
```
6. All files needed for the further steps are now in the folders `./sp/filtered_bam/` (`.bam` files) and `./sp/MACS2/` (`.narrowPeak` file as the peak file).
